File: apps/ios/Packages/TrackEvolutionKit/Sources/TrackEvolutionKit/Models/Wrapped.swift
import Foundation

/// `GET /api/wrapped/:year` — Season Wrapped (NS-36): one calendar year's
/// numbers, past events only, computed on the server by `src/lib/wrapped.ts`.
///
/// **No screen reads this yet.** Wrapped is web-first; the model is here so the
/// golden contract decodes, and so a native Wrapped next November is this model
/// plus a screen — the computation is the server's, so there is nothing to port.
/// Every card is optional because the server skips a card with no data rather
/// than sending it empty.
public struct Wrapped: Codable, Hashable, Sendable {
    public var year: Int
    /// Every year with a past event, newest first — the picker.
    public var years: [Int]
    /// Today, ISO `YYYY-MM-DD`, while the year is running; nil once it has ended.
    public var through: String?
    public var name: String?
    public var totals: WrappedTotals
    public var mostDriven: WrappedMostDriven?
    public var improvement: WrappedImprovement?
    public var fastest: WrappedFastest?
    public var newTracks: [WrappedTrack]
    public var hottest: WrappedHottest?
    /// The one tier-dependent field: nil for a free account (the client draws
    /// the two Pro cards locked); for Pro, each card is nil when there is no data.
    /// Absent altogether from the public share (`GET /api/share/:slug/wrapped/:year`),
    /// which carries the free card set only and decodes into this same model.
    public var pro: WrappedPro?

    public enum CodingKeys: String, CodingKey {
        case year, years, through, name, totals, improvement, fastest, hottest, pro
        case mostDriven = "most_driven"
        case newTracks = "new_tracks"
    }
}

public struct WrappedTotals: Codable, Hashable, Sendable {
    public var events: Int
    /// Fractional, like `Event.days`.
    public var trackDays: Double
    public var tracks: Int
    public var laps: Int
    public var hours: Double
    public var miles: Double
    /// How many of `tracks` had a known lap length — "across N of M tracks".
    public var milesTracksCounted: Int

    public enum CodingKeys: String, CodingKey {
        case events, tracks, laps, hours, miles
        case trackDays = "track_days"
        case milesTracksCounted = "miles_tracks_counted"
    }
}

public struct WrappedTrack: Codable, Hashable, Sendable {
    public var trackId: Int
    public var trackName: String

    public enum CodingKeys: String, CodingKey {
        case trackId = "track_id"
        case trackName = "track_name"
    }
}

public struct WrappedMostDriven: Codable, Hashable, Sendable {
    public var trackId: Int
    public var trackName: String
    public var trackDays: Double
    public var laps: Int
    public var bestMs: Int?

    public enum CodingKeys: String, CodingKey {
        case laps
        case trackId = "track_id"
        case trackName = "track_name"
        case trackDays = "track_days"
        case bestMs = "best_ms"
    }
}

public struct WrappedImprovement: Codable, Hashable, Sendable {
    public var trackId: Int
    public var trackName: String
    public var bestBefore: Int
    public var bestThisYear: Int
    public var gainMs: Int
    /// `"prior_years"`, or `"first_event"` for a first year at the track.
    public var baseline: String

    public enum CodingKeys: String, CodingKey {
        case baseline
        case trackId = "track_id"
        case trackName = "track_name"
        case bestBefore = "best_before"
        case bestThisYear = "best_this_year"
        case gainMs = "gain_ms"
    }
}

public struct WrappedFastest: Codable, Hashable, Sendable {
    public var trackId: Int
    public var trackName: String
    public var bestMs: Int
    public var eventId: Int
    public var date: String

    public enum CodingKeys: String, CodingKey {
        case date
        case trackId = "track_id"
        case trackName = "track_name"
        case bestMs = "best_ms"
        case eventId = "event_id"
    }
}

public struct WrappedHottest: Codable, Hashable, Sendable {
    public var eventId: Int
    public var trackName: String
    public var date: String
    public var tempC: Double

    public enum CodingKeys: String, CodingKey {
        case date
        case eventId = "event_id"
        case trackName = "track_name"
        case tempC = "temp_c"
    }
}

public struct WrappedPro: Codable, Hashable, Sendable {
    public var tire: WrappedTire?
    public var topSpeed: WrappedTopSpeed?

    public enum CodingKeys: String, CodingKey {
        case tire
        case topSpeed = "top_speed"
    }
}

/// The tyre with the most track days on it this year (garage consumables).
public struct WrappedTire: Codable, Hashable, Sendable {
    public var partId: Int
    public var vehicleId: Int
    public var vehicleName: String
    public var name: String
    public var trackDays: Double
    public var hours: Double

    public enum CodingKeys: String, CodingKey {
        case name, hours
        case partId = "part_id"
        case vehicleId = "vehicle_id"
        case vehicleName = "vehicle_name"
        case trackDays = "track_days"
    }
}

/// The year's highest stored speed sample, km/h.
public struct WrappedTopSpeed: Codable, Hashable, Sendable {
    public var kph: Double
    public var trackId: Int
    public var trackName: String
    public var eventId: Int
    public var date: String

    public enum CodingKeys: String, CodingKey {
        case kph, date
        case trackId = "track_id"
        case trackName = "track_name"
        case eventId = "event_id"
    }
}

File: apps/android/core/src/main/kotlin/app/trackevolution/core/model/Wrapped.kt
package app.trackevolution.core.model

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

/**
 * `GET /api/wrapped/:year` — Season Wrapped (NS-36): one calendar year's
 * numbers, past events only, computed on the server by `src/lib/wrapped.ts`.
 *
 * Read by `WrappedScreen` in `:app`, through [app.trackevolution.core.WrappedStory]'s
 * card rules. The computation is the server's, so nothing here is ported but the
 * presentation. Every card is nullable because the server skips a card with no
 * data rather than sending it empty.
 */
@Serializable
public data class Wrapped(
    val year: Int,
    /** Every year with a past event, newest first — the picker. */
    val years: List<Int>,
    /** Today, ISO `YYYY-MM-DD`, while the year is running; null once it has ended. */
    val through: String? = null,
    val name: String? = null,
    val totals: WrappedTotals,
    @SerialName("most_driven") val mostDriven: WrappedMostDriven? = null,
    val improvement: WrappedImprovement? = null,
    val fastest: WrappedFastest? = null,
    @SerialName("new_tracks") val newTracks: List<WrappedTrack>,
    val hottest: WrappedHottest? = null,
    /**
     * The one tier-dependent field: null for a free account (the client draws
     * the two Pro cards locked); for Pro, each card is null when there is no data.
     * Absent altogether from the public share (`GET /api/share/:slug/wrapped/:year`),
     * which carries the free card set only and decodes into this same model.
     */
    val pro: WrappedPro? = null,
)

@Serializable
public data class WrappedTotals(
    val events: Int,
    /** Fractional, like [Event.days]. */
    @SerialName("track_days") val trackDays: Double,
    val tracks: Int,
    val laps: Int,
    val hours: Double,
    val miles: Double,
    /** How many of [tracks] had a known lap length — "across N of M tracks". */
    @SerialName("miles_tracks_counted") val milesTracksCounted: Int,
)

@Serializable
public data class WrappedTrack(
    @SerialName("track_id") val trackId: Int,
    @SerialName("track_name") val trackName: String,
)

@Serializable
public data class WrappedMostDriven(
    @SerialName("track_id") val trackId: Int,
    @SerialName("track_name") val trackName: String,
    @SerialName("track_days") val trackDays: Double,
    val laps: Int,
    @SerialName("best_ms") val bestMs: Int? = null,
)

@Serializable
public data class WrappedImprovement(
    @SerialName("track_id") val trackId: Int,
    @SerialName("track_name") val trackName: String,
    @SerialName("best_before") val bestBefore: Int,
    @SerialName("best_this_year") val bestThisYear: Int,
    @SerialName("gain_ms") val gainMs: Int,
    /** `"prior_years"`, or `"first_event"` for a first year at the track. */
    val baseline: String,
)

@Serializable
public data class WrappedFastest(
    @SerialName("track_id") val trackId: Int,
    @SerialName("track_name") val trackName: String,
    @SerialName("best_ms") val bestMs: Int,
    @SerialName("event_id") val eventId: Int,
    val date: String,
)

@Serializable
public data class WrappedHottest(
    @SerialName("event_id") val eventId: Int,
    @SerialName("track_name") val trackName: String,
    val date: String,
    @SerialName("temp_c") val tempC: Double,
)

@Serializable
public data class WrappedPro(
    val tire: WrappedTire? = null,
    @SerialName("top_speed") val topSpeed: WrappedTopSpeed? = null,
)

/** The tyre with the most track days on it this year (garage consumables). */
@Serializable
public data class WrappedTire(
    @SerialName("part_id") val partId: Int,
    @SerialName("vehicle_id") val vehicleId: Int,
    @SerialName("vehicle_name") val vehicleName: String,
    val name: String,
    @SerialName("track_days") val trackDays: Double,
    val hours: Double,
)

/** The year's highest stored speed sample, km/h. */
@Serializable
public data class WrappedTopSpeed(
    val kph: Double,
    @SerialName("track_id") val trackId: Int,
    @SerialName("track_name") val trackName: String,
    @SerialName("event_id") val eventId: Int,
    val date: String,
)
